### ISSUE TITLE
lxc-copy fails to create an LVM snapshot when running in parallel

### ISSUE THREAD

**Issue by BenoitKnecht** [lxc/lxc#2220]
# Required information

 * Distribution: Ubuntu
 * Distribution version: Xenial (16.04)
 * The output of
   * `lxc-copy --version`: 2.0.8
   * `uname -a`: Linux controller-01 4.4.0-116-generic #140-Ubuntu SMP Mon Feb 12 21:23:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

# Issue description

I'm trying to create LXC containers using the LVM backend, by cloning a base container called `ubuntu-xenial-amd64`. The `lxc` volume group contains the following logical volume:

```console
root@controller-01:~# lvs
  LV                                                          VG   Attr       LSize Pool Origin              Data%  Meta%  Move Log Cpy%Sync Convert
  ubuntu-xenial-amd64                                         lxc  owi-a-s--- 5.00g                                                                 
```

Creating containers based on that volume sequentially works fine, but trying to create several containers in parallel fails with:

```
lxc-copy: bdev/bdev.c: mount_unknown_fs: 780 failed to determine fs type for '/dev/lxc/ubuntu-xenial-amd64'
lxc-copy: bdev/bdev.c: detect_fs: 675 failed mounting /dev/lxc/ubuntu-xenial-amd64 onto /var/lib/lxc/ubuntu-xenial-amd64/rootfs to detect fstype
lxc-copy: bdev/bdev.c: detect_fs: 653 child exited early - fstype not found
lxc-copy: bdev/bdev.c: bdev_copy: 388 failed getting pathnames for cloned storage: /dev/lxc/ubuntu-xenial-amd64
lxc-copy: lxccontainer.c: copy_storage: 2865 Error copying storage.
clone failed
```

# Steps to reproduce

 1. `lxc-copy --snapshot -B lvm --name ubuntu-xenial-amd64 --newname test0 -L 5G --logfile test0.log --logpriority DEBUG& lxc-copy --snapshot -B lvm --name ubuntu-xenial-amd64 --newname test1 -L 5G --logfile test1.log --logpriority DEBUG`
2. Only one of the two containers is created. The logs for both containers are attached below.

The problem seems to be that `lxc-copy` attempts to mount `/dev/lxc/ubuntu-xenial-amd64` onto `/var/lib/lxc/ubuntu-xenial-amd64/rootfs` to detect its filesystem, but if running in parallel, it's already mounted there, so the attempt fails.

As far as I can tell, the logic in `detect_fs()` hasn't changed between 2.0.8 and master. I'm guessing that a possible way to solve this issue would be to mount `/dev/lxc/ubuntu-xenial-amd64` in a unique, random location, but I haven't tried implementing that yet.

# Information to attach

- [test1.log](https://github.com/lxc/lxc/files/1814851/test1.log)
- [test0.log](https://github.com/lxc/lxc/files/1814852/test0.log)

**Comment by brauner** [comment]
Or not fail when it is already mounted.

**Comment by BenoitKnecht** [comment]
> Or not fail when it is already mounted.

But you might get another race condition if you do that, if the device is umounted by the first `lxc-copy` process before you had the chance to determine the filesystem type, right?

**Comment by brauner** [comment]
Yeah, definitely. The random mount is something one could do. I wouldn't nack a patch
/cc @hallyn?

**Comment by BenoitKnecht** [comment]
Thinking about this a bit more, I was wondering if it would make sense to get the filesystem type from udev, instead of attempting to mount the device.

The filesystem type is exposed by udev as the `ID_FS_TYPE` property.

**Comment by hallyn** [comment]
Seems like using a tempdir to do the mount into would avoid the race.

I like the udev idea:

serge@sergeh2:~$ udevadm info /dev/vg0/irc | grep FS_TYPE
E: ID_FS_TYPE=ext4

Is this something you'd want to write a patch for?

**Comment by BenoitKnecht** [comment]
> Is this something you'd want to write a patch for?

Sure, I'll give it a shot!

Thanks.
